### ISSUE TITLE
Implement changing refresh rates based on the video's framerate

### ISSUE THREAD

**Issue by Dudemanguy911** [mpv-player/mpv#552]
I think this would be a pretty neat thing to have. I'd imagine it'd be nice if mpv could read whatever the framerate of the video is (24p, 30p, whatever) and then use xrandr or something to detect the available refresh rates you have at that resolution and then automatically switch the monitor to the appropriate, integer-multiple rate. So say you have a 1920x1080 monitor with modelines that allow it to be at 60hz or 72hz. Playing a 30p video would switch it to 60hz and then playing a 24p video would switch it to 72hz.

**Comment by lvml** [comment]
Since this issue is still open I just wanted to advertise that the functionality requested is at least available as a plugin script at https://github.com/lvml/mpv-plugin-xrandr

Probably couldn't hurt to have it built into mpv, though ;-)

**Comment by wm4** [comment]
>Probably couldn't hurt to have it built into mpv

Undecided whether I really want this. The implementation would probably be relatively similar, except the VO would support switching rates directly.

**Comment by wp9015362** [comment]
Any update on this one?

In case it might matter:

Kodi (http://kodi.tv/) does have this feature.

It's under System -> Settings -> Video -> Playback.

Here's the decription from the Wiki (http://kodi.wiki/view/Settings/Videos#Playback):

> **Adjust display refresh rate to match video**
> Defines when the refresh rate adjustments should take place.
> Activate to automatically set the refresh rate which is best matched with the video that is playing (default off). This potentially allows for perfectly smooth video reproduction, as video material can be recorded in a variety of frame rates (23.976fps, 24fps, 25fps, 29.97fps, 30 fps, etc.) which need to be properly matched by the refresh rate of the display, in order to be displayed smoothly.

You can set the option to "Off" or to "Always" or to "On start / stop".

When it is set to "Off", then the refresh rate is not being adjusted.

When it is set to "Always", then it constantly adjusts the display refresh rate as far as I know (if the video is variable frame rate or something like that).

If it is set to "On start / stop", then it only adjusts the display refresh rate at the beginning of the video and returns to the default refresh rate once the video has been stopped.

There is also another neat suboption: 

> **Pause during refresh rate change**
> Inserts a pause for the computer to adjust refresh rate. Sometimes needed to prevent things like slight stuttering when a video starts. Range 0.1 milliseconds to 20 seconds. (default off)

This option is also good for if your monitor takes some time until it display an image after a refresh rate change. So, let's say if your monitor shows a blank image for two seconds while switching refresh rates, then you simply set the option to anywhere between 2 or 3 seconds and Kodi will pause the video for 2-3 seconds while the refresh rate change occurs, so you don't miss anything from the video playback :smiley:.

It would be very nice if mpv would also have such features :wink:.

Any chance it could be implemented like in Kodi?

Regards

**Comment by wm4** [comment]
I don't have any plans for this. There are some scripts which do this.

**Comment by lvml** [comment]
BTW: The above mentioned script https://github.com/lvml/mpv-plugin-xrandr just got the additional feature to switch back to the original rate when leaving mpv.
Alas, using the external xrandr tool means we can learn about the refresh rate used before invoking mpv only with the "rounded" precision of the xrandr output.

**Comment by Nojevah** [comment]
Sorry, I just wanted to vote for this request, hoping it would motivate someone.
On Windows, there's no way to change refresh rate depending on the screen. For a TV, using 23Hz refresh rate is clearly better than 60Hz with (blurry) interpolation.

**Comment by haasn** [comment]
@Nojevah On windows there's the `autospeedwin` lua script.

**Comment by Nojevah** [comment]
Yes I've tried it. But there are many problems:
- You can't specify a setting by display. For example, I want to change refresh rate only for TV, not monitor and both are "monitor 0" (I'm using single monitor). We could hope NirCmd allowed to choose Display ID, but it's not the case or it's not working. I've tried to email him/them, without answer unfortunately.
- The script choose "nearest" frequencies, So for a 23.976fps movie, it prefers 24Hz refresh rate instead of 23 (23 being 23,976 in drivers).

Tbh, I come from MPC+madVR where both offer option , so it's a big miss for me.

**Comment by garoto** [comment]
See if this app [ ChangeScreenResolution](http://tools.taubenkorb.at/change-screen-resolution/) works best. If it does, then kevinlekiller's script can probably be adapted to use it instead or nircmdc.

**Comment by Nojevah** [comment]
Nice tool indeed. Still I have the same problems:
- You can't specify a setting by display. You can only choose display 0 or 1. In general, you don't want to change refresh rate on your monitor, it's only when you play a video on TV that you want to adapt refresh rate. Yet it's possible to get monitor ID.
- script still needs to be modified, and I can't do it.

**Comment by tamodolo** [comment]
Some screens don't support 24Hz but do support the next multiple: 48Hz (most PC monitors actually do support. Test yours if you want.). So make this feature works in a way that you can choose the actual refresh rate based on source video. Like 2x source: 24 plays at 48Hz and 30 plays at 60. Higher speeds like 50 and 60fps plays at native refresh rate as they are widely suported. Or make a config file were you tell the player what to do.

**Comment by zc62** [comment]
@Nojevah The second problem in `autospeedwin` should have gone away since my PR was merged (https://github.com/kevinlekiller/mpv_scripts/commit/5001b049b032b921fc3c4de7e8f40fa6f159368e) in November 2017.

I originally thought this feature is useful, but later someone asked me what if you watch vfr (variable frame rate) contents. Also, playing with following settings on a 60.00Hz monitor perfectly resemble the behavior of a 23.976fps movie played on a 23.976Hz TV:
```
video-sync=display-resample
video-sync-max-video-change=5
interpolation=yes
tscale=oversample
```
So basically a 23.976fps movie will be sped up by 0.1% to 24.00fps, then by repeating the frames and blending nearest frames, the motion judder is removed. (See https://github.com/mpv-player/mpv/wiki/Interpolation.) The good thing is, even if you have multiple displays, none of them need special configurations. Just interpolation. Also, it saves the time that the screen goes black when you adjust the refresh rate.

**Comment by Dudemanguy911** [comment]
Dear lord, this is still open? Scripts solved it long time ago imo. Woops.

**Comment by Nojevah** [comment]
It is still an issue, obviously.
But as long as no dev is interested, you can indeed close this. If someone is interested, he'll submit a patch.

**Comment by lvml** [comment]
BTW: "The script that solved it" for Linux some time ago moved to a new home at https://gitlab.com/lvml/mpv-plugin-xrandr

**Comment by ValZapod** [comment]
> basically a 23.976fps movie will be sped up by 0.1% to 24.00fps

What??? Is this a joke? No, 24.000 should be 24.000 and 24000/1001 must be 24000/1001

**Comment by 07416** [comment]
> [2019] Dear lord, this is still open? Scripts solved it long time ago imo. 

@Dudemanguy Is there a Wayland-compatible script?

**Comment by Dudemanguy** [comment]
Not to my knowledge since this differs depending on the compositor. You can easily make one for sway and/or wlroots-based compositors though.

**Comment by 07416** [comment]
there's a script for KDE on Wayland: https://gitlab.com/smaniottonicola/mpv-kscreen-doctor

I don't see others with Wayland compatibility on [the wiki](https://github.com/mpv-player/mpv/wiki/User-Scripts).

**Comment by SavantGarde** [comment]
> Dear lord, this is still open? Scripts solved it long time ago imo. Woops.

Not for macOS.

**Comment by gw711** [comment]
Windows script?

**Comment by kasper93** [comment]
> Windows script?

https://github.com/kevinlekiller/mpv_scripts/tree/master/autospeedwin
